### PR TITLE
blockchain: Make ver 5 to 6 db upgrades work again.

### PR DESCRIPTION
**Pull request by davecgh**
This allows version 5 to 6 database upgrades to work again by implementing code that is independent of the main code and works with the data structures as they existed in the version 5 database.

While here, it also updates some other cases in both older and newer upgrade code to make the upgrades stable even if the primary code changes.

An high-level overview of the key changes is:

- Clear the failed validation status of the v2 block index entries   directly instead of loading the code into the main index, modifying it, and flushing it
- Avoid using the unversioned main funcs related to serializing and deserializing the spend journal
- Implement code to directly create a script source from a v2 spend journal entry
- Reimplement the migration code that initializes the v2 GCS filters to independently read the block tree and discover the main chain

### CODE REVIEW

File: blockchain/upgrade.go [hunk 1]
```diff
@@ -942,17 +1099,222 @@ func (s scriptSource) PrevScript(prevOut *wire.OutPoint) (uint16, []byte, bool)
...
+			// Deserialize the script version.
+			if offset >= len(serialized) {
+				str := "unexpected end of data after flags"
+				return nil, errDeserialize(str)
+			}
+			scriptVersion, bytesRead := deserializeVLQ(serialized[offset:])
+			offset += bytesRead
+
+			// Decode the compressed script size and ensure there are enough
+			// bytes left in the slice for it
```
Comment (matheusd) [inline blockchain/upgrade.go:1266]: Missing full stop at the end.

**Review by matheusd — APPROVED** [review]
Tested by syncing a few testnet blocks using a v1.4.0 binary, then performing the migration.

I also spot checked a few filters by running `getcfilterv2` on the upgraded node and on new nodes.

Just found two small typos.

File: blockchain/upgrade.go (hunk 1 above)
Comment (davecgh) [reply blockchain/upgrade.go:1266]: Fixed.

File: blockchain/upgrade.go [hunk 2]
```diff
@@ -942,17 +1099,222 @@ func (s scriptSource) PrevScript(prevOut *wire.OutPoint) (uint16, []byte, bool)
...
+}
+
+// scriptSourceFromSpendJournalV1 uses the legacy v1 spend journal along with
+// the provided block to create a source of previous transaction scripts and
+// versions spent by the block.
+func scriptSourceFromSpendJournalV1(dbTx database.Tx, block *wire.MsgBlock) (scriptSource, error) {
+	// Load the serialized spend journal entry from the database, construct the
+	// full list of transactions that spend outputs (notice the coinbase
+	// transaction is excluded since it can't spend anything), and perform an
+	// intiial sanity check to ensure there is serialized data for the block
```
[the remark this answers is not in the document]
Comment (davecgh) [reply blockchain/upgrade.go:1154]: Updated.

**Review by dnldd — APPROVED** [review]

**Review by matheusd — APPROVED** [review]

**Pull request merged**